cli/init.ml

```diff
@@ -0,0 +1,109 @@
+  Common.Logs.app "Calculating Git repositories to vendor";
+  compute_repos ~packages >>= fun repos ->
+  let duniverse = { Duniverse.root_packages; excludes; pins; remotes; branch; packages; repos } in
+  let file = Fpath.(repo // Config.duniverse_file) in
+  Duniverse.save ~file duniverse >>= fun () ->
+  Common.Logs.app "Wrote duniverse file with %a entries to %a."
+    Fmt.(styled `Green int)
+    (List.length repos)
+    Fmt.(styled `Cyan Fpath.pp)
+    file;
+  Ok ()
+
+open Cmdliner
+
+let explicit_root_packages =
+  let doc =
+    "opam packages to calculate duniverse for. If not supplied, any local opam metadata files are \
+     used as the default package list."
+  in
+  Arg.(value & pos_all string [] & info [] ~doc ~docv:"PACKAGES")
+
+let excludes =
+  let doc =
+    "Packages to exclude from the output list. You can use this to remove the root packages so \
+     they are not duplicated in the vendor directory.  Repeat this flag multiple times for more \
+     than one exclusion."
+  in
+  Arg.(value & opt_all string [] & info [ "exclude"; "x" ] ~docv:"EXCLUDE" ~doc)
+
+let pins =
+  let open Types.Opam in
+  let doc =
+    "Packages to pin for the latest opam metadata and source. You can separate the package name \
+     and a url and a remote branch via commas to specify a manual url (e.g. \
+     $(i,mirage,git://github.com/avsm/mirage,fixme)).  If a url is not specified then the \
+     $(i,--dev) pin is used.  If a branch is not specified then the default remote branch is \
+     used. Repeat this flag multiple times for more than one exclusion."
+  in
+  let fin s =
+    match Astring.String.cuts ~sep:"," s with
+    | [] -> failwith "unexpected pin error"
+    | [ pin ] -> Ok { pin; url = None; tag = None }
+    | [ pin; url ] -> Ok { pin; url = Some url; tag = None }
+    | [ pin; url; tag ] -> Ok { pin; url = Some url; tag = Some tag }
+    | _ -> failwith "pins must have maximum of 3 commas"
+  in
+  let fout ppf { pin; url; tag } =
+    match (url, tag) with
+    | None, _ -> Fmt.(pf ppf "%s" pin)
+    | Some url, None -> Fmt.(pf ppf "%s,%s" pin url)
+    | Some url, Some tag -> Fmt.(pf ppf "%s,%s,%s" pin url tag)
+  in
+  let t = Arg.conv ~docv:"PIN" (fin, fout) in
+  Arg.(value & opt_all t [] & info [ "pin"; "p" ] ~docv:"PIN" ~doc)
+
+let remotes =
+  let doc =
+    "Extra opam remotes to add when resolving package names. Repeat this flag multiple times for \
+     more than one remote."
+  in
+  Arg.(value & opt_all string [] & info [ "opam-remote" ] ~docv:"OPAM_REMOTE" ~doc)
+
+let info =
+  let exits = Term.default_exits in
+  let doc = "analyse opam files to generate an initial $(b,duniverse.sxp)" in
+  let man = [] in
+  Term.info "init" ~doc ~exits ~man
+
+let term =
+  let open Term in
+  term_result
+    ( const run $ Common.Arg.repo $ Common.Arg.branch $ explicit_root_packages $ excludes $ pins
+    $ remotes $ Common.Arg.setup_logs () )
+
+let cmd = (term, info)
```

Comment:
Why a new library ?
Half of `Cmdliner` code is still in `bin/duniverse.ml`, which contains only that plus `let () = Term.exit ...`
I'd personally prefer if `Cmdliner` code was in `bin/` and logic (eg. `Init.run`) in `lib/`.
This library is meant to be opened so I think the module names should be longer, eg. `cli_common.ml` and `init_cmd.ml`.